### PR TITLE
Fix dist error with lr decay layer

### LINKED ISSUE #9429: using learning rate decay in optimizer in dist train cause run sub program error enforce  framework::product(lr_dims) == 1 failed, 0 != 1
```
E0328 03:54:46.540556  3693 listen_and_serv_op.cc:152] run sub program error enforce framework::product(lr_dims) == 1 failed, 0 != 1
Learning rate should have 1 element at [/paddle/paddle/fluid/operators/sgd_op.cc:36]
PaddlePaddle Call Stacks:
0       0x7f6a2a6ddf1cp paddle::platform::EnforceNotMet::EnforceNotMet(std::__exception_ptr::exception_ptr, char const*, int) + 572
1       0x7f6a2ae144eap paddle::operators::SGDOp::InferShape(paddle::framework::InferShapeContext*) const + 970
2       0x7f6a2aed7358p paddle::framework::OperatorWithKernel::RunImpl(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) const + 104
3       0x7f6a2aed4cd8p paddle::framework::OperatorBase::Run(paddle::framework::Scope const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) + 72
4       0x7f6a2a78ff66p paddle::framework::Executor::RunPreparedContext(paddle::framework::ExecutorPrepareContext*, paddle::framework::Scope*, bool, bool) + 1750
5       0x7f6a2a790e77p paddle::framework::Executor::Run(paddle::framework::ProgramDesc const&, paddle::framework::Scope*, int, bool, bool) + 103
```

### PR DESCRIPTION
**Pull request by Yancey1989**
Fixed #9429

### CODE REVIEW

File: python/paddle/fluid/optimizer.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 from regularizer import append_regularization_ops
 from clip import append_gradient_clip_ops, error_clip_callback
 from contextlib import contextmanager
+from distribute_transpiler import UnionFind
```
Comment (typhoonzero) [inline python/paddle/fluid/optimizer.py]: `optimizer.py` should not depend on distribute_transpiler, either put `_get_lr_decay_ops` in the transpiler or put `UnionFind` in a single file. I'd prefer the first method, because we don't have to change current demo files then.

Comment (Yancey1989) [reply python/paddle/fluid/optimizer.py]: > optimizer.py should not depend on distribute_transpiler

Thanks @typhoonzero ,I think it's a good point, and maybe we can pass `Optimizer` instance to `transpile` interface so that we can support regularization for future.

Comment (typhoonzero) [reply python/paddle/fluid/optimizer.py]: We can consider that when moving regularizer and clipping to the server side.

Comment (Yancey1989) [reply python/paddle/fluid/optimizer.py]: Done, moved to transpiler.

File: python/paddle/fluid/distribute_transpiler.py [hunk 2]
```diff
@@ -786,3 +795,32 @@ def _get_output_map_from_op(self, varmap, op):
...
+        # find lr variables by optimize op
+        lr_vars = set()
+        for op in self.optimize_ops:
+            if self._is_opt_op(op):
+                lr_vars.add(op.input("LearningRate")[0])
+
+        find_ops = []
+        # find ops which output is lr var
+        # make a union-find struct by all ops
+        block = default_main_program().global_block()
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply python/paddle/fluid/distribute_transpiler.py]: Done.
